File: 103.二叉树的锯齿形层次遍历.cpp
/*
 * @lc app=leetcode.cn id=103 lang=cpp
 *
 * [103] 二叉树的锯齿形层次遍历
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> zigzagLevelOrder(TreeNode* root) {
        vector<vector<int>> res;
        if(!root) return res;
        deque<TreeNode*> nodes;
        nodes.push_back(root);
        bool order=true;
        TreeNode* tmp;
        while(!nodes.empty()){
            int s=nodes.size();
            vector<int> floor;
            while(s){
                if(order){
                    tmp=nodes.front();
                    nodes.pop_front();
                    if(tmp->left) nodes.push_back(tmp->left);
                    if(tmp->right) nodes.push_back(tmp->right);
                }else{
                    tmp=nodes.back();
                    nodes.pop_back();
                    if(tmp->right) nodes.push_front(tmp->right);
                    if(tmp->left) nodes.push_front(tmp->left);
                }
                floor.push_back(tmp->val);
                s--;
            }
            order=!order;
            res.push_back(floor);
        }
        return res;
    }
};
// @lc code=end

File: 103.二叉树的锯齿形层次遍历.py
#
# @lc app=leetcode.cn id=103 lang=python3
#
# [103] 二叉树的锯齿形层次遍历
#

# @lc code=start
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def zigzagLevelOrder(self, root: TreeNode) -> List[List[int]]:
        res=[]
        if root is None: return res
        nodes=deque()
        nodes.append(root)
        order=True
        tmp=TreeNode()
        while len(nodes)!=0:
            s=len(nodes)
            floor=[]
            while s!=0:
                if order:
                    tmp=nodes.popleft()
                    if tmp.left: nodes.append(tmp.left)
                    if tmp.right: nodes.append(tmp.right)
                else:
                    tmp=nodes.pop()
                    if tmp.right: nodes.appendleft(tmp.right)
                    if tmp.left: nodes.appendleft(tmp.left)
                floor.append(tmp.val)
                s-=1
            order=not order
            res.append(floor)
        return res
# @lc code=end

